elytron/WFCORE-4058-automatic-keystore-detection.adoc

```diff
@@ -0,0 +1,92 @@
+= Automatic detection of file based KeyStore types.
+:author:            Maritn Mazanek
+:email:             mmazanek@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+* If a user does not specify the type of a file based KeyStore can we auto detect the type?
+
+* There is most likely a finite set of types we would want to automatically recognise, possibly we could inspect the
+contents ourselves to detect different types - alternative attempt initialisation of different KeyStore objects until one succeeds. With the latter we would need to be careful one doesn't silently succeed when it should have failed.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/EAP7-675[EAP7-675]
+* https://issues.jboss.org/browse/ELY-1650[ELY-1650]
+* https://issues.jboss.org/browse/WFCORE-4058[WFCORE-4058]
+
+//=== Related Issues
+
+//* https://issues.jboss.org/browse/WFLY[WFLY-XXXX]
+
+=== Dev Contacts
+
+* mailto:mmazanek@redhat.com[Martin Mazanek]
+
+=== QE Contacts
+
+* TBD
+
+=== Affected Projects or Components
+
+* WildFly, Security
+
+== Requirements
+
+=== Hard Requirements
+
+* Point of this change is to be able to load keystore without specifying its type.
+
+== Implementation Plan
+
+* Finite set of supported keystore types for automatic detection: JKS, JCEKS, PKCS12, BKS, BCFKS, UBER
+* Some types are fairly easily recognizable, as they start with magic number: JKS (0xfeedfeed), JCEKS (0xcececece) - although if a keystore starts with such number it doesn't necessarily mean that it is that type
+* Detect types of encrypted keystores? They may be a good candidate for try until success (let the provider do the decryption for us)
+
+* The idea is to prioritize different techniques with a priority
+** Magic number - check if the store contains one of known magic numbers - if yes, then try to load it and see if it is really that keystore
+** File extension - try to parse as a keystore based on the file extension
+** Try remaining supported types in a specific order / don't try anything and fail
+
+* Recognizable patterns:
```

Comment:
A more suitable question would be - what mechanism to the existing providers use to verify if the file is suitable for them to load?  If they are also using these same signature tests it will not be possible for them to be changed in the future.

However if we use the signature checks to adjust the priority order for a brute force check a subsequent change changes very little.

Reply:
The providers do similar checks - with keystores that have magic numbers they check that, they usualy have version number after that, so they checkt that and the rest is just parsing, so if something is corrupted, it will simply fail parsing (+ there is usually some kind of digest at the end). Others don't perform any verification at all, they start parsing and if the data is not in the right format then you have parsing fail.

Anyways the JKS, JCEKS and PKCS12 should not change its recognizable patterns, as JKS and JCEKS have magic numbers (which definitely won't change) and PKCS12 is purely ASN.1 DER data, which will always start with SEQUENCE tag 0x30

Reply:
What could change is the version numbers of the keystores, but that can be left out for the providers to figure out, as we can figure out the type without it